flux 0.31.3: source-controller panics when we are using a HelmRepository with an OCI registry without specifying the type

Hi everyone,

I’ve configured our private helm repository (Artifact registry from Google) to deploy our applications from a helmrelease.

I’m currently facing to the following error when I try to reconcile our HelmRepository

```
failed to fetch Helm repository index: failed to cache index to temporary file: object required
````
When I check the flux pods, I can see that the source-controller stop working and when I'm looking for the logs I have a `panic`

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1d818f7]

goroutine 496 [running]:
github.com/fluxcd/source-controller/controllers.(*HelmChartReconciler).buildFromHelmRepository(0xc0006fc800, {0x2a55c58, 0xc0005a3830}, 0xc005850fc0, 0xc002955a40, 0xc002167140)
        github.com/fluxcd/source-controller/controllers/helmchart_controller.go:573 +0x1317
github.com/fluxcd/source-controller/controllers.(*HelmChartReconciler).reconcileSource(0xc0006fc800, {0x2a55c58, 0xc0005a3830}, 0xc005850fc0, 0xc002167140)
        github.com/fluxcd/source-controller/controllers/helmchart_controller.go:441 +0x7e7
github.com/fluxcd/source-controller/controllers.(*HelmChartReconciler).reconcile(0x2ac5360, {0x2a55c58, 0xc0005a3830}, 0xc005850fc0, {0xc000679bf0, 0x3, 0xc00072e430})
        github.com/fluxcd/source-controller/controllers/helmchart_controller.go:281 +0x1a4
github.com/fluxcd/source-controller/controllers.(*HelmChartReconciler).Reconcile(0xc0006fc800, {0x2a55c58, 0xc0005a3350}, {{{0xc00072e430, 0x25d5f60}, {0xc000446de0, 0x30}}})
        github.com/fluxcd/source-controller/controllers/helmchart_controller.go:259 +0x5b0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc00045cbb0, {0x2a55c58, 0xc0005a3260}, {{{0xc00072e430, 0x25d5f60}, {0xc000446de0, 0x415454}}})
        sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114 +0x26f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00045cbb0, {0x2a55bb0, 0xc000bc3e80}, {0x245dc80, 0xc0003a6600})
        sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311 +0x33e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00045cbb0, {0x2a55bb0, 0xc000bc3e80})
        sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:223 +0x357
```

Here you can find my HelmRepository

```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: HelmRepository
metadata:
  annotations:
    reconcile.fluxcd.io/requestedAt: "2022-07-13T15:35:17.939321+02:00"
  labels:
    kustomize.toolkit.fluxcd.io/name: configs
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: helm-charts-registry
  namespace: flux-system
spec:
  interval: 10m
  secretRef:
    name: shared-helm-registry
  timeout: 60s
  url: oci://europe-west4-docker.pkg.dev/xxxxxxxxxxxxxx/helm-charts
```

I'm using flux `0.31.3`

--- a/controllers/helmchart_controller.go
+++ b/controllers/helmchart_controller.go
@@ -392,7 +392,8 @@ func (r *HelmChartReconciler) reconcileSource(ctx context.Context, obj *sourcev1
 
 	// Assert source has an artifact
 	if s.GetArtifact() == nil || !r.Storage.ArtifactExist(*s.GetArtifact()) {
-		if helmRepo, ok := s.(*sourcev1.HelmRepository); !ok || !helmreg.IsOCI(helmRepo.Spec.URL) {
+		// Set the condition to indicate that the source has no artifact for all types except OCI HelmRepository
+		if helmRepo, ok := s.(*sourcev1.HelmRepository); !ok || helmRepo.Spec.Type != sourcev1.HelmRepositoryTypeOCI {
 			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, "NoSourceArtifact",
 				"no artifact available for %s source '%s'", obj.Spec.SourceRef.Kind, obj.Spec.SourceRef.Name)
 			r.eventLogf(ctx, obj, events.EventTypeTrace, "NoSourceArtifact",

--- a/controllers/helmchart_controller_test.go
+++ b/controllers/helmchart_controller_test.go
@@ -515,6 +515,156 @@ func TestHelmChartReconciler_reconcileSource(t *testing.T) {
 	}
 }
 
+func TestHelmChartReconciler_reconcileFromHelmRepository(t *testing.T) {
+	g := NewWithT(t)
+
+	const (
+		chartName          = "helmchart"
+		chartVersion       = "0.2.0"
+		higherChartVersion = "0.3.0"
+		chartPath          = "testdata/charts/helmchart"
+	)
+
+	serverFactory, err := helmtestserver.NewTempHelmServer()
+	g.Expect(err).NotTo(HaveOccurred())
+	defer os.RemoveAll(serverFactory.Root())
+
+	for _, ver := range []string{chartVersion, higherChartVersion} {
+		g.Expect(serverFactory.PackageChartWithVersion(chartPath, ver)).To(Succeed())
+	}
+	g.Expect(serverFactory.GenerateIndex()).To(Succeed())
+
+	tests := []struct {
+		name       string
+		beforeFunc func(repository *sourcev1.HelmRepository)
+		assertFunc func(g *WithT, obj *sourcev1.HelmChart)
+	}{
+		{
+			name: "Reconciles chart build",
+			assertFunc: func(g *WithT, obj *sourcev1.HelmChart) {
+				key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+				// Wait for HelmChart to be Ready
+				g.Eventually(func() bool {
+					if err := testEnv.Get(ctx, key, obj); err != nil {
+						return false
+					}
+					if !conditions.IsReady(obj) || obj.Status.Artifact == nil {
+						return false
+					}
+					readyCondition := conditions.Get(obj, meta.ReadyCondition)
+					return obj.Generation == readyCondition.ObservedGeneration &&
+						obj.Generation == obj.Status.ObservedGeneration
+				}, timeout).Should(BeTrue())
+
+				// Check if the object status is valid.
+				condns := &status.Conditions{NegativePolarity: helmChartReadyCondition.NegativePolarity}
+				checker := status.NewChecker(testEnv.Client, condns)
+				checker.CheckErr(ctx, obj)
+			},
+		},
+		{
+			name: "Stalling on invalid repository URL",
+			beforeFunc: func(repository *sourcev1.HelmRepository) {
+				repository.Spec.URL = "://unsupported" // Invalid URL
+			},
+			assertFunc: func(g *WithT, obj *sourcev1.HelmChart) {
+				key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+				// Wait for HelmChart to be FetchFailed == true
+				g.Eventually(func() bool {
+					if err := testEnv.Get(ctx, key, obj); err != nil {
+						return false
+					}
+					if !conditions.IsTrue(obj, sourcev1.FetchFailedCondition) {
+						return false
+					}
+					// observedGeneration is -1 because we have no successful reconciliation
+					return obj.Status.ObservedGeneration == -1
+				}, timeout).Should(BeTrue())
+
+				// Check if the object status is valid.
+				condns := &status.Conditions{NegativePolarity: helmChartReadyCondition.NegativePolarity}
+				checker := status.NewChecker(testEnv.Client, condns)
+				checker.CheckErr(ctx, obj)
+			},
+		},
+		{
+			name: "Stalling on invalid oci repository URL",
+			beforeFunc: func(repository *sourcev1.HelmRepository) {
+				repository.Spec.URL = strings.Replace(repository.Spec.URL, "http", "oci", 1)
+			},
+			assertFunc: func(g *WithT, obj *sourcev1.HelmChart) {
+				key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+				// Wait for HelmChart to be Ready
+				g.Eventually(func() bool {
+					if err := testEnv.Get(ctx, key, obj); err != nil {
+						return false
+					}
+					if !conditions.IsTrue(obj, sourcev1.FetchFailedCondition) {
+						return false
+					}
+					// observedGeneration is -1 because we have no successful reconciliation
+					return obj.Status.ObservedGeneration == -1
+				}, timeout).Should(BeTrue())
+
+				// Check if the object status is valid.
+				condns := &status.Conditions{NegativePolarity: helmChartReadyCondition.NegativePolarity}
+				checker := status.NewChecker(testEnv.Client, condns)
+				checker.CheckErr(ctx, obj)
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			server := testserver.NewHTTPServer(serverFactory.Root())
+			server.Start()
+			defer server.Stop()
+
+			ns, err := testEnv.CreateNamespace(ctx, "helmchart")
+			g.Expect(err).ToNot(HaveOccurred())
+			defer func() { g.Expect(testEnv.Delete(ctx, ns)).To(Succeed()) }()
+
+			repository := sourcev1.HelmRepository{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "helmrepository-",
+					Namespace:    ns.Name,
+				},
+				Spec: sourcev1.HelmRepositorySpec{
+					URL: server.URL(),
+				},
+			}
+
+			if tt.beforeFunc != nil {
+				tt.beforeFunc(&repository)
+			}
+
+			g.Expect(testEnv.CreateAndWait(ctx, &repository)).To(Succeed())
+
+			obj := sourcev1.HelmChart{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "helmrepository-reconcile-",
+					Namespace:    ns.Name,
+				},
+				Spec: sourcev1.HelmChartSpec{
+					Chart:   chartName,
+					Version: chartVersion,
+					SourceRef: sourcev1.LocalHelmChartSourceReference{
+						Kind: sourcev1.HelmRepositoryKind,
+						Name: repository.Name,
+					},
+				},
+			}
+			g.Expect(testEnv.Create(ctx, &obj)).To(Succeed())
+
+			if tt.assertFunc != nil {
+				tt.assertFunc(g, &obj)
+			}
+		})
+	}
+}
+
 func TestHelmChartReconciler_buildFromHelmRepository(t *testing.T) {
 	g := NewWithT(t)
 

